docs/heuristics.rst

```diff
@@ -84,4 +84,54 @@ or::
     def grouping(files, dcmfilter, seqinfo):
         seqinfos = collections.OrderedDict()
         ...
-        return seqinfos  # ordered dict containing seqinfo objects: list of DICOMs
\ No newline at end of file
+        return seqinfos  # ordered dict containing seqinfo objects: list of DICOMs
+
+
+-------------------------------
+``POPULATE_INTENDED_FOR_OPTS``
+-------------------------------
+
+Dictionary to specify options to populate the ``'IntendedFor'`` field of the ``fmap``
+jsons.
+
+When a BIDS session has ``fmaps``, they can automatically be assigned to be used for
+susceptibility distortion correction of other non-``fmap`` images in the session
+(populating the ``'IntendedFor'`` field in the ``fmap`` json file).
+
+For this automated assignment, ``fmaps`` are taken as groups (``_phase`` and ``_phasediff``
+images and the corresponding ``_magnitude`` images; consecutive Spin-Echo images collected
+with opposite phase encoding polarity (``pepolar`` case); etc.).
+
+This is achieved by checking, for every non-``fmap`` image in the session, which ``fmap``
+groups are suitable candidates to correct for distortions in that image.  Then, if there is
+more than one candidate (e.g., if there was a ``fmap`` collected at the beginning of the
+session and another one at the end), the user can specify which one to use.
+
+The parameters that can be specified and the allowed options are defined in ``bids.py``:
+ - ``'matching_parameter'``: The imaging parameter that needs to match between the ``fmap``
+   and an image for the ``fmap`` to be considered as a suitable to correct that image.
+   Allowed options are:
+
+   * ``'Shims'``: ``heudiconv`` will check the ``ShimSetting`` in the ``.json`` files and
+     will only assign ``fmaps`` to images if the ``ShimSettings`` are identical for both.
+   * ``'ImagingVolume'``: both ``fmaps`` and images will need to have the same the imaging
+     volume (the header affine transformation: position, orientation and voxel size, as well
+     as number of voxels along each dimensions).
+
+
+ - ``'criterion'``: Criterion to decide which of the candidate ``fmaps`` will be assigned to
+   a given file, if there are more than one. Allowed values are:
+
+   * ``'First'``: The first matching ``fmap``.
+   * ``'Closest'``: The closest in time to the beginning of the image acquisition.
+
+.. note::
+  Example::
+
+    POPULATE_INTENDED_FOR_OPTS = {
+            'matching_parameter': 'ImagingVolume',
+            'criterion': 'Closest'
+    }
+
+If ``POPULATE_INTENDED_FOR_OPTS`` is not present in the heuristic file, ``IntendedFor``
+will not be populated automatically.
```

Comment:
hm,  I think it is not the case currently -- I do not spot any condition at  https://github.com/nipy/heudiconv/pull/482/files#diff-d19d089741b3844c199c8236ec1c0fcd43400b2ce53cd587d203255add863c39R546 or around, and when I tried now with v0.9.0-97-g723fa67 an `-f reproin` it happily populated it.

I am ok either way.  But if to demand it to be defined, make check for it to be None and not just empty dict, so empty dict would instruct "enable but use default opts".

Also I wondered if we should make `matching_parameter` into `matching_parameters` and default to have both zooms and sizes to be matched.  This way we have a better chance of doing "the right thing". Also it would allow later to expand with additional ones we have discussed at some point.

Reply:
> I am ok either way. But if to demand it to be defined, make check for it to be None and not just empty dict, so empty dict would instruct "enable but use default opts".

It would be [here](https://github.com/cbinyu/heudiconv/blob/723fa67673a6b504da9a3b2a0b2e09814615305f/heudiconv/convert.py#L213) right?

Reply:
> Also I wondered if we should make `matching_parameter` into `matching_parameters` and default to have both zooms and sizes to be matched. This way we have a better chance of doing "the right thing". Also it would allow later to expand with additional ones we have discussed at some point.

You mean, have `matching_parameters` to be a list and have all elements in the list to match?
Sure, I can do that.

Reply:
> > I am ok either way. But if to demand it to be defined, make check for it to be None and not just empty dict, so empty dict would instruct "enable but use default opts".
> 
> It would be [here](https://github.com/cbinyu/heudiconv/blob/723fa67673a6b504da9a3b2a0b2e09814615305f/heudiconv/convert.py#L213) right?

yes, there it would default to `None`, not `{}` and then somewhere [there in convert](https://github.com/cbinyu/heudiconv/blob/723fa67673a6b504da9a3b2a0b2e09814615305f/heudiconv/convert.py#L527) would do `if populate_intended_for_opts is not None:` for the subsequent block of code which deals with triggering `populate_intended_for`

Reply:
> > Also I wondered if we should make `matching_parameter` into `matching_parameters` and default to have both zooms and sizes to be matched. This way we have a better chance of doing "the right thing". Also it would allow later to expand with additional ones we have discussed at some point.
> 
> You mean, have `matching_parameters` to be a list and have all elements in the list to match?
> Sure, I can do that.

yes.  We better do it now so later we do not need to deal with handling both `matching_parameter` and `matching_parameters`.

PS I am so looking forward to have this PR merged! ;) Thank you for all the work and bearing with all my comments/wishes

Reply:
OK. I think both issues have been addressed in https://github.com/nipy/heudiconv/pull/482/commits/5a06034e4ed0e2e52ca8e446b1384bf92aa40e60 and https://github.com/nipy/heudiconv/pull/482/commits/ccb96dfdce52a6376a42d47b07951882570d044a.
I also [made `ImagingVolume` the default](https://github.com/nipy/heudiconv/pull/482/commits/b72ecd8f5b701a3898dc67a6930d6fe499275998) for `populate_intended_for` (that means that the position, orientation, resolution and numbers of voxels should match).